Sources/CredentialsFacebook/TypeSafeFacebookToken.swift

```diff
@@ -0,0 +1,163 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+
+
+import Kitura
+import KituraNet
+import LoggerAPI
+import Credentials
+import Foundation
+import TypeDecoder
+
+/// A protocol that a user's type can conform to representing a user authenticated using a
+/// Facebook OAuth token.
+/**
+ ### Usage Example: ###
+ public struct ExampleProfile: TypeSafeFacebookToken {
+    static var appID: String = "yourFacebookAppID"
+
+    let id: String
+
+    let name: String
+ 
+    let email: String?
+ }
+ router.get("/facebookProfile") { (user: ExampleProfile, respondWith: (ExampleProfile?, RequestError?) -> Void) in
+    respondWith(user, nil)
+ }
+ */
+public protocol TypeSafeFacebookToken: TypeSafeFacebook {
+
+    // MARK: Instance fields
+
+    /// The application-scoped ID field. Note that this field uniquely identifies a user
+    /// wihin the context of the application represented by the token.
+    var id: String { get }
+
+    /// The subject's display name.
+    var name: String { get }
+
+}
+
+/// The cache element for keeping facebook profile information.
+private class FacebookCacheElement {
+    /// The user profile information stored as `TypeSafeFacebookToken`.
+    internal var userProfile: TypeSafeFacebookToken
+
+    /// Initialize a `FacebookCacheElement`.
+    ///
+    /// - Parameter profile: the `TypeSafeFacebookToken` to store.
+    internal init (profile: TypeSafeFacebookToken) {
+        userProfile = profile
+    }
+}
+
+// An internal type to hold the mapping from a user's type to an appropriate token cache.
+//
+// It is a workaround for the inability to define stored properties in a protocol extension.
+//
+// We use the `debugDescription` of the user's type (via `String(reflecting:)`) as the
+// dictionary key.
+private struct TypeSafeFacebookTokenCache {
+    internal static var cacheForType: [String: NSCache<NSString, FacebookCacheElement>] = [:]
+}
+
+extension TypeSafeFacebookToken {
+
+    // Associates a profile cache with the user's type. This relieves the user from having to
+    // declare a usersCache property on their conforming type.
+    private static var usersCache: NSCache<NSString, FacebookCacheElement> {
+        let key = String(reflecting: Self.self)
+        if let usersCache = TypeSafeFacebookTokenCache.cacheForType[key] {
+            return usersCache
+        } else {
+            let usersCache = NSCache<NSString, FacebookCacheElement>()
+            TypeSafeFacebookTokenCache.cacheForType[key] = usersCache
+            return usersCache
+        }
+    }
+
+    /// Authenticate incoming request using Facebook OAuth token.
```

Comment:
It reads as though it needs a few more words - Authenticate "the/an" incoming request using "the" Facebook OAuth token. ?